File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/null_object/lab/variable_flyweight_factory.cpp
// variable_flyweight_factory.cpp

#include <string>
#include "dp_map.h"
#include "variable.h"
#include "variable_flyweight_factory.h"

variable_flyweight_factory * variable_flyweight_factory::the_instance(nullptr);

variable * variable_flyweight_factory::get_flyweight_for(const std::string & a_name)
{
	if (!my_map.contains(a_name))
	{
		my_map.put(a_name, new variable(a_name));
	}

	variable * result = my_map.get(a_name);

	return result;
}

variable_flyweight_factory::variable_flyweight_factory()
{
}

variable_flyweight_factory & variable_flyweight_factory::get_instance()
{
	if (the_instance == nullptr)
	{
		the_instance = new variable_flyweight_factory();
	}

	return *the_instance;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/interpreter/lab/VariableFlyweightFactory.java
// VariableFlyweightFactory.java

public class VariableFlyweightFactory
{
	public Variable getFlyweightFor(String aName)
	{
		if (!myMap.contains(aName))
		{
			myMap.put(aName, new Variable(aName));
		}

		Variable result = myMap.get(aName);

		return result;
	}

	private DpMap<String, Variable> myMap;

	private VariableFlyweightFactory()
	{
		myMap = new DpMap<String, Variable>();
	}

	public static VariableFlyweightFactory getInstance()
	{
		if (theInstance == null)
		{
			theInstance = new VariableFlyweightFactory();
		}

		return theInstance;
	}

	private static VariableFlyweightFactory theInstance = null;
}
